transport/transport_test.go

```diff
@@ -0,0 +1,262 @@
+		Request: &csr.CertificateRequest{
+			CN: "localhost test certificate",
+		},
+		Profiles: map[string]map[string]string{
+			"paths": map[string]string{
+				"private_key": testKey,
+				"certificate": testCert,
+			},
+			"cfssl": {
+				"label":   testLabel,
+				"profile": testProfile,
+				"remote":  testRemote,
+			},
+		},
+	}
+)
+
+func TestTransportSetup(t *testing.T) {
+	var before = 55 * time.Second
+	var err error
+
+	tr, err = NewTransport(before, testIdentity)
+	if err != nil {
+		t.Fatalf("failed to set up transport: %v", err)
+	}
+}
+
+func TestRefreshKeys(t *testing.T) {
+	err := tr.RefreshKeys()
+	if err != nil {
+		t.Fatalf("%v", err)
+	}
+}
+
+func TestAutoUpdate(t *testing.T) {
+	// To force a refresh, make sure that the certificate is
+	// updated 5 seconds from now.
+	cert := tr.Provider.Certificate()
+	if cert == nil {
+		t.Fatal("no certificate from provider")
+	}
+
+	certUpdates := make(chan time.Time, 0)
+	errUpdates := make(chan error, 0)
+	oldBefore := tr.Before
+	before := cert.NotAfter.Sub(time.Now())
+	before -= 5 * time.Second
+	tr.Before = before
+	defer func() {
+		tr.Before = oldBefore
+		PollInterval = 30 * time.Second
+	}()
+
+	PollInterval = 2 * time.Second
+
+	go tr.AutoUpdate(certUpdates, errUpdates)
+	log.Debugf("waiting for certificate update or error from auto updater")
+	select {
+	case <-certUpdates:
+		// Nothing needs to be done
+	case err := <-errUpdates:
+		t.Fatalf("%v", err)
+	case <-time.After(15 * time.Second):
+		t.Fatal("timeout waiting for update")
+	}
+}
+
+var (
+	l             *Listener
+	testLKey      = "testdata/server.key"
+	testLCert     = "testdata/server.pem"
+	testLIdentity = &core.Identity{
+		Request: &csr.CertificateRequest{
+			CN:    "localhost test certificate",
+			Hosts: []string{"127.0.0.1"},
+		},
+		Profiles: map[string]map[string]string{
+			"paths": map[string]string{
+				"private_key": testLKey,
+				"certificate": testLCert,
+			},
+			"cfssl": {
+				"label":   testLabel,
+				"profile": testProfile,
+				"remote":  testRemote,
+			},
+		},
+	}
+)
+
+func testListen(t *testing.T) {
+	log.Debug("listener waiting for connection")
+	conn, err := l.Accept()
+	if err != nil {
+		t.Fatalf("%v", err)
+	}
+
+	log.Debugf("client has connected")
+	conn.Write([]byte("hello"))
+
+	conn.Close()
+}
+
+func TestListener(t *testing.T) {
+	var before = 55 * time.Second
+
+	trl, err := NewTransport(before, testLIdentity)
+	if err != nil {
+		t.Fatalf("failed to set up transport: %v", err)
+	}
+
+	trl.Identity.Request.CN = "localhost test server"
+
+	err = trl.RefreshKeys()
+	if err != nil {
+		t.Fatalf("%v", err)
+	}
+
+	core.SystemRoots = x509.NewCertPool()
+
+	caCert, err := tr.CA.CACertificate()
+	if err != nil {
+		t.Fatalf("%v", err)
+	}
+	if !core.SystemRoots.AppendCertsFromPEM(caCert) {
+		t.Fatal("no certificates could be added to system roots")
+	}
+
+	core.SystemRoots.AddCert(tr.Provider.Certificate())
+
+	l, err = Listen("127.0.0.1:8765", trl)
```

Comment:
This is pretty slick, you can easily set up an HTTPS server with s.Serve(*l).

It would be nice to add an instruction manual for people who have set up their HTTP(S) endpoints with ListenAndServe(TLS) to switch to the new transport.